text/0000-clippy-uno.md

```diff
@@ -0,0 +1,671 @@
+- [option_map_unit_fn](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#option_map_unit_fn): Checks for usage of `option.map(f)` where f is a function
+or closure that returns the unit type.
+- [overflow_check_conditional](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#overflow_check_conditional): Detects classic underflow/overflow checks.
+- [transmute_ptr_to_ref](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#transmute_ptr_to_ref): Checks for transmutes from a pointer to a reference.
+- [chars_next_cmp](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#chars_next_cmp): Checks for usage of `.chars().next()` on a `str` to check
+if it starts with a given char.
+- [transmute_bytes_to_str](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#transmute_bytes_to_str): Checks for transmutes from a `&[u8]` to a `&str`.
+- [identity_conversion](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#identity_conversion): Checks for always-identical `Into`/`From` conversions.
+- [double_parens](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#double_parens): Checks for unnecessary double parentheses.
+- [zero_divided_by_zero](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#zero_divided_by_zero): Checks for `0.0 / 0.0`.
+- [useless_asref](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_asref): Checks for usage of `.as_ref()` or `.as_mut()` where the
+types before and after the call are the same.
+- [too_many_arguments](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#too_many_arguments): Checks for functions with too many parameters.
+- [range_zip_with_len](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#range_zip_with_len): Checks for zipping a collection with the range of
+`0.._.len()`.
+- [temporary_assignment](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#temporary_assignment): Checks for construction of a structure or tuple just to
+assign a value in it.
+- [no_effect](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#no_effect): Checks for statements which have no effect.
+- [short_circuit_statement](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#short_circuit_statement): Checks for the use of short circuit boolean conditions as
+a
+statement.
+- [cast_lossless](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#cast_lossless): Checks for on casts between numerical types that may
+be replaced by safe conversion functions.
+- [unnecessary_operation](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#unnecessary_operation): Checks for expression statements that can be reduced to a
+sub-expression.
+- [cyclomatic_complexity](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#cyclomatic_complexity): Checks for methods with high cyclomatic complexity.
```

Comment:
Cyclomatic complexity is one of the best examples of Goodhart's law, meaning that high cyclomatic complexity might be a sign of hard-to-understand code, but intentionally trying to reduce it is almost certainly going to make it even worse. As someone else once pointed out to me, you can reduce cyclomatic complexity to 1 by Church-encoding all conditionals and replacing all loops by recursion. Which does not improve readability.

So I think this lint should be `allow` by default.

Reply:
> replacing all loops by recursion. Which does not improve readability.

This statement probably makes a lot of haskellers very angry :stuck_out_tongue: 

On a more serious note:

There is almost always a bad way to resolve any lint. The correct solution to the CC lint triggering in case there is no way to nicely refactor the function is to increase the limit for the function.

What I have noticed is that high CC functions can often be split up into multiple ones which, instead of having everything in scope, just have a few arguments and often just one or two return values. Increasing modularity is a good thing in my book. Ideally the CC lint would be able to suggest good spots for refactoring out code into separate functions.

Note that the CC lint is very Rust-tuned and not just the straight forward original implementation. This is due to the fact that Rust uses `match` extensively (which in vanilla CC has a CC of the number of arms + the CC of all arm bodies). Other languages usually introduce a single dynamic dispatch call (which has a CC of 1). Thus the clippy-CC treats a `match` as having a CC of just its arms' bodies CC. So if all the arm bodies have a CC of 1, the entire match has a CC of 2 (to account for *some* branching going on).

The same goes for short circuting operations (`&&` and `||`), these do not increase the CC.

Reply:
> This statement probably makes a lot of haskellers very angry 😛

More like: ![Not angry, just disappointed ;)](https://memegenerator.net/img/instances/64210492/im-not-angry-im-just-disappointed.jpg)

Tho, once you've done some manual recursion a few times, you start using folds and maps and stuff. There are many ways to be a Haskell programmer.. See: [The Evolution of a Haskell Programmer](https://www.willamette.edu/~fruehr/haskell/evolution.html).

I think your observation is spot on otherwise. Using more higher order functions is also a nice way to get rid of rightward drift and cyclomatic complexity, or a more compositional style in general. The worst thing you can have is long and deep functions; Long and shallow functions are not the worst, but also not very great. Short and shallow gives you nice readability.

Reply:
> The worst thing you can have is long and deep functions; Long and shallow functions are not the worst, but also not very great. Short and shallow gives you nice readability.

This is sometimes true, but not universally. I'm a Haskell programmer too, and I've seen some absolutely unreadable short and shallow functions that overused higher-order functions and point-free style, whose readability would have greatly improved by not code-golfing and being overly clever.

Which is really my point. CC is an incredibly simplistic heuristic for something that is inherently difficult to measure, and trying to optimise your code specifically for it will make it worse, rather than better. The programmer needs to consider trade-offs when it comes to readability, for instance the trade-off between rightward-drift and extra indirection. This requires good judgment. I'd really prefer if the allow-by-default lints would stick to recommendations that make your code unambiguously better (like 'use a byte literal rather than casting char literals to bytes').

Reply:
> I'd really prefer if the allow-by-default lints would stick to recommendations that make your code unambiguously better

Did you mean warn-by-default? (cyclomatic is warn but I see a good argument here for making it allow in the pedantic group)

Reply:
@Manishearth I did mean warn-by-default, thanks for pointing that out!

Reply:
> This is sometimes true, but not universally.

Of course. :) Very few things are universally true. In this case I'd say it is more often true than not.

> I'm a Haskell programmer too,

👋 ❤️

> I've seen some absolutely unreadable short and shallow functions that overused higher-order functions and point-free style, whose readability would have greatly improved by not code-golfing and being overly clever.

Sure; if you write code looking like it was processed through http://pointfree.io/, then it will look unreadable. Particularly if you combine `flip` and sectioning of `(.)`. However, using function composition linearly (where associativity applies) to a bunch of small functions makes code both more maintainable, reusable and readable. How readable such code is also depends on how many combinators you have in your toolbox. Of course, you need to strike a balance between pointfree style and pointfull style. A few well chosen places to use a lambda here and there *can* improve readability. And while the Haskeller who has entered their [rigorous phase](https://terrytao.wordpress.com/career-advice/theres-more-to-mathematics-than-rigour-and-proofs/) (according to themselves), and who have started to grasp the more advanced category-theoretical concepts, might eschew `do` notation, a post-rigorous Haskeller will probably enjoy using it.

Something I find particularly nice about shallow and short functions is that you also get more type signatures (assuming you put type signatures on all top level functions, which you should), which helps understanding (due to free theorems) in my opinion. Functions also become more readily testable with `QuickCheck` thus improving the reliability of your code as a consequence. I also personally tend to document every single top level function I write with haddock in Haskell. As a consequence, even if you don't understand the details of the small function, you can more easily grasp the overall semantics and the big picture. Having the habit of writing short functions also is a boon for recognizing patterns and extracting out common algorithms.

If you want to compare styles, here is a [code style document](https://github.com/DATX02-17-26/DATX02-17-26/blob/dev/docs/code-style.md) I wrote for my BsC thesis project.

Contrast this to long and deep functions with a bunch of nested `where` blocks and `case` expressions. Then you get fewer type signatures, logic that is harder to test, and less reuse.

Also, if we consider a language such as Agda or Idris, splitting proof obligations out to a different function can make development much easier by having fewer things to deal with in your short term memory. Edwin Brady demonstrates this in his talk on [Type-driven Development with Idris](https://www.youtube.com/watch?v=gonVdPyVwQQ)

> Which is really my point. CC is an incredibly simplistic heuristic for something that is inherently difficult to measure, and trying to optimize your code specifically for it will make it worse, rather than better.

While being a simple heuristic as you say, I think it often hits well. In particular, my view is that a lot of Rust code already tends to involve too many long and deep functions (spanning 300 lines) and very long files that make it hard (*at least for me*) to gauge the overall picture. I think modules, as a consequence, are too underused in Rust.  As such, I would much prefer to retain the *warn-by-default* nature of the `cyclomatic_complexity` lint in the complexity group.

> The programmer needs to consider trade-offs when it comes to readability, for instance the trade-off between rightward-drift and extra indirection. This requires good judgment.

It does. But I find that it also depends on how well the out-extracted function is named. A small and short function calling other aptly named small and short functions can read like pseudo-code. A good IDE can also enable you to quickly jump to the definition of the helper functions. Therefore, I think the default setting should be to write short and sweet definitions.

> I'd really prefer if the warn-by-default lints would stick to recommendations that make your code unambiguously better (like 'use a byte literal rather than casting char literals to bytes').

For the reasons I've outlined above, I disagree. To me, lints such as `too_many_arguments`, `type_complexity`, and `cyclomatic_complexity` induce good habits, even if I've been hit by them myself and felt like it was unwarranted.
I also

Reply:
>> This is sometimes true, but not universally.

> Of course. :) Very few things are universally true. In this case I'd say it is more often true than not.

Most suggestions in this lint category actually are always or nearly always improvements. I can't think of any reason to ever write `if x == true` or `x + 0` or `if c { true } else {false }` or `'c' as u8`, while this lint (as well as the two others you mention, which I also don't think should warn by default) may improve the code or make it worse, depending on the situation and personal taste.

I'm not saying this is a bad lint or that high cyclomatic complexity is great, which your reply seems to be arguing against. I'm just saying it doesn't belong together with the other cut-and-dried 'here are clear and simple instructions to make your code better' lints.

Reply:
The same reasoning applies to all lints with thresholds. Type complexity and number of single char bindings have arbitrary but sensible thresholds from which on they'll tell you to do "better". In super-mathy code, many single char bindings might be fine. Or you have a very deeply nested type that is obvious due to many different nicely named types.

If you encounter a situation where the threshold does not fit your use case, increase it. If you regress even further, clippy will tell you again with the new threshold. I created the `complexity` lint group exactly for these kinds of lints, but at some point figured that other things like `x == true` belonged here, too, because they are also things that are "too complex" for some notion of "too".

I agree that there are two different kinds of lints in the `complexity` group, but I also don't see how to reasonably split them up further (without just splitting out all the thresholded lints). Categorization always comes with a little "error" because otherwise you'd end up with as many categories as you have stuff to categorize.

Reply:
@fhartwig I agree with @oli-obk, but I would also say that I would expect these lints to be in a group called "complexity". I also think that cyclomatic complexity (at some arbitrary good default) warnings are nearly always (but not 100% of the time) improvements. Or enough of the time at least.